syntax = "proto3";

package talos.icd.v1;

import "google/protobuf/empty.proto";

// The historical transport exposed a command wrapper with message and command
// identifiers. gRPC already correlates each request/response pair based on the
// RPC method and stream, so the explicit MessageId/CommandId fields are no
// longer required and have been omitted here.

service TalosOperator {
	rpc Handshake(HandshakeRequest) returns (HandshakeResponse);
	rpc PolarPanDiscrete(PolarPanDiscreteRequest) returns (CommandResult);
	rpc Home(HomeRequest) returns (CommandResult);
	rpc PolarPanContinuousStart(PolarPanContinuousStartRequest) returns (CommandResult);
	rpc PolarPanContinuousStop(google.protobuf.Empty) returns (CommandResult);
	rpc CartesianMoveDiscrete(CartesianMoveDiscreteRequest) returns (CommandResult);
	rpc CartesianMoveContinuousStart(CartesianMoveContinuousStartRequest) returns (CommandResult);
	rpc CartesianMoveContinuousStop(google.protobuf.Empty) returns (CommandResult);
	rpc ExecuteHardwareOperation(ExecuteHardwareOperationRequest) returns (ExecuteHardwareOperationResponse);
	rpc GetSpeed(google.protobuf.Empty) returns (GetSpeedResponse);
	rpc SetSpeed(SetSpeedRequest) returns (CommandResult);
	rpc SavePosition(SavePositionRequest) returns (CommandResult);
	rpc DeletePosition(DeletePositionRequest) returns (CommandResult);
	rpc GoToPosition(GoToPositionRequest) returns (CommandResult);
	rpc SetPolarPosition(SetPolarPositionRequest) returns (CommandResult);
	rpc GetPolarPosition(GetPolarPositionRequest) returns (GetPolarPositionResponse);
	rpc SetCartesianPosition(SetCartesianPositionRequest) returns (CommandResult);
	rpc GetCartesianPosition(GetCartesianPositionRequest) returns (GetCartesianPositionResponse);
}

message CommandResult {
	// Matches the UINT16 return codes from the ICD; stored in a wider type to
	// follow protobuf scalar options.
	uint32 return_code = 1;
}

message HandshakeRequest {
	// Operator/Arm model connection identifier (UINT16 in the ICD).
	uint32 operator_id = 1;
}

message HandshakeResponse {
	CommandResult status = 1;
	uint32 major_version = 2;
	uint32 minor_version = 3;
}

message PolarPanDiscreteRequest {
	int32 delta_azimuth_tenths_deg = 1;
	int32 delta_altitude_tenths_deg = 2;
	uint32 delay_ms = 3;
	uint32 execution_time_ms = 4;
}

message HomeRequest {
	uint32 delay_ms = 1;
}

message PolarPanContinuousStartRequest {
	// Each axis indicator is constrained to -1, 0, or 1.
	int32 moving_azimuth = 1;
	int32 moving_altitude = 2;
}

message CartesianMoveDiscreteRequest {
	int32 delta_x_tenths_mm = 1;
	int32 delta_y_tenths_mm = 2;
	int32 delta_z_tenths_mm = 3;
	uint32 delay_ms = 4;
	uint32 execution_time_ms = 5;
}

message CartesianMoveContinuousStartRequest {
	// Each axis indicator is constrained to -1, 0, or 1.
	int32 moving_x = 1;
	int32 moving_y = 2;
	int32 moving_z = 3;
}

message ExecuteHardwareOperationRequest {
	uint32 subcommand_id = 1;
	bytes payload = 2;
}

message ExecuteHardwareOperationResponse {
	CommandResult status = 1;
	uint32 subcommand_id = 2;
	bytes payload = 3;
}

message GetSpeedResponse {
	CommandResult status = 1;
	uint32 speed = 2;
}

message SetSpeedRequest {
	uint32 speed = 1;
}

message SavePositionRequest {
	string name = 1;
	bool anchor = 2;
	// When empty, the parent reference is ignored just like the legacy ICD.
	string parent_reference = 3;
}

message DeletePositionRequest {
	string name = 1;
}

message GoToPositionRequest {
	string name = 1;
}

message SetPolarPositionRequest {
	string name = 1;
	int32 delta_tenths_deg = 2;
	int32 azimuth_tenths_deg = 3;
	int32 radius_tenths = 4;
}

message GetPolarPositionRequest {
	string name = 1;
}

message GetPolarPositionResponse {
	CommandResult status = 1;
	int32 delta_tenths_deg = 2;
	int32 azimuth_tenths_deg = 3;
	int32 radius_tenths = 4;
}

message SetCartesianPositionRequest {
	string name = 1;
	int32 x_tenths_mm = 2;
	int32 y_tenths_mm = 3;
	int32 z_tenths_mm = 4;
}

message GetCartesianPositionRequest {
	string name = 1;
}

message GetCartesianPositionResponse {
	CommandResult status = 1;
	int32 x_tenths_mm = 2;
	int32 y_tenths_mm = 3;
	int32 z_tenths_mm = 4;
}

